crates/wasmtime/src/runtime/gc/noextern.rs

```diff
@@ -0,0 +1,188 @@
+///                 ;; If the reference is null, return.
+///                 local.get 0
+///                 ref.is_null noextern
+///                 br_if 0
+///
+///                 ;; If the reference was not null (which is impossible)
+///                 ;; then raise a trap.
+///                 unreachable
+///             )
+///         )
+///     "#,
+/// )?;
+///
+/// let mut store = Store::new(&engine, ());
+/// let instance = Instance::new(&mut store, &module, &[])?;
+/// let f = instance.get_func(&mut store, "f").unwrap();
+///
+/// // We can cast a `(ref null noextern)`-taking function into a typed function that
+/// // takes an `Option<NoExtern>` via the `Func::typed` method.
+/// let f = f.typed::<Option<NoExtern>, ()>(&store)?;
+///
+/// // We can call the typed function, passing the null `noextern` reference.
+/// let result = f.call(&mut store, NoExtern::null());
+///
+/// // The function should not have trapped, because the reference we gave it was
+/// // null (as it had to be, since `NoExtern` is uninhabited).
+/// assert!(result.is_ok());
+/// # Ok(())
+/// # }
+/// ```
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub struct NoExtern {
+    _inner: Uninhabited,
+}
+
+impl NoExtern {
+    /// Get the null `(ref null noextern)` (aka `nullexternref`) reference.
+    #[inline]
+    pub fn null() -> Option<Self> {
+        None
+    }
+
+    /// Get the null `(ref null noextern)` (aka `nullexternref`) reference as a
+    /// [`Ref`].
+    #[inline]
+    pub fn null_ref() -> Ref {
+        Ref::Extern(None)
+    }
+
+    /// Get the null `(ref null noextern)` (aka `nullexternref`) reference as a
+    /// [`Val`].
+    #[inline]
+    pub fn null_val() -> Val {
+        Val::ExternRef(None)
+    }
+}
+
+unsafe impl WasmTy for NoExtern {
+    type Abi = NoExtern;
+
+    #[inline]
+    fn valtype() -> ValType {
+        ValType::Ref(RefType::new(false, HeapType::NoExtern))
+    }
+
+    #[inline]
+    fn compatible_with_store(&self, _store: &StoreOpaque) -> bool {
+        match self._inner {}
+    }
+
+    #[inline]
+    fn dynamic_concrete_type_check(&self, _: &StoreOpaque, _: bool, _: &HeapType) -> Result<()> {
+        match self._inner {}
+    }
+
+    #[inline]
+    fn is_non_i31_gc_ref(&self) -> bool {
+        match self._inner {}
+    }
+
+    #[inline]
+    unsafe fn abi_from_raw(_raw: *mut ValRaw) -> Self::Abi {
+        unreachable!("NoExtern is uninhabited")
+    }
+
+    #[inline]
+    unsafe fn abi_into_raw(_abi: Self::Abi, _raw: *mut ValRaw) {
+        unreachable!("NoExtern is uninhabited")
+    }
+
+    #[inline]
+    fn into_abi(self, _store: &mut AutoAssertNoGc<'_>) -> Result<Self::Abi> {
+        unreachable!("NoExtern is uninhabited")
+    }
+
+    #[inline]
+    unsafe fn from_abi(_abi: Self::Abi, _store: &mut AutoAssertNoGc<'_>) -> Self {
+        unreachable!("NoExtern is uninhabited")
+    }
+}
+
+unsafe impl WasmTy for Option<NoExtern> {
+    type Abi = *mut NoExtern;
```

Comment:
Is this perhaps best modelled as `usize`? Because technically this is a pointer-sized integer that is a relative offset into the GC heap?

Reply:
Yeah it should be a `u32`, since if inhabited `noextern` would be a `VMGcRef` which is a `NonZeroU32` so if we make it optional/nullable we are left with a `u32`.